feat: Replications and WAF alignment for App Store - `avm/res/app-configuration/configuration-store`

JFolberth opened the pull request:
## Description

- Added replicas
- Update settings per pester results
- ReadMes Update
- Configured replicas for WAF alignment
- Removed orphan Markdown

Fixes #1502
Closes #1502
-->

## Pipeline Reference

<!-- Insert your Pipeline Status Badge below -->

| Pipeline |
| -------- |
| [![avm.res.app-configuration.configuration-store](https://github.com/JFolberth/bicep-registry-modules/actions/workflows/avm.res.app-configuration.configuration-store.yml/badge.svg)](https://github.com/JFolberth/bicep-registry-modules/actions/workflows/avm.res.app-configuration.configuration-store.yml)|

## Type of Change

<!-- Use the check-boxes [x] on the options that are relevant. -->

- [ ] Update to CI Environment or utlities (Non-module effecting changes)
- [x] Azure Verified Module updates:
  - [ ] Bugfix containing backwards compatible bug fixes, and I have NOT bumped the MAJOR or MINOR version in `version.json`:
    - [ ] Someone has opened a bug report issue, and I have included "Closes #{bug_report_issue_number}" in the PR description.
    - [ ] The bug was found by the module author, and no one has opened an issue to report it yet.
  - [x] Feature update backwards compatible feature updates, and I have bumped the MINOR version in `version.json`.
  - [ ] Breaking changes and I have bumped the MAJOR version in `version.json`.
  - [x] Update to documentation

## Checklist

- [x] I'm sure there are no other open Pull Requests for the same update/change
- [x] I have run `Set-AVMModule` locally to generate the supporting module files.
- [x] My corresponding pipelines / checks run clean and green without any errors or warnings

<!--  Please keep up to day with the contribution guide at https://aka.ms/avm/contribute/bicep -->

JFolberth commented:
@matebarabas @eriqua 

Looking for feedback on this on two items.

I was not sure on replicas what to do with locations...I found avm\res\aad\domain-service\tests\e2e\waf-aligned\main.test.bicep where locations are hard coded and figured maybe this is how it should be handled?

I also per https://azure.github.io/PSRule.Rules.Azure/en/rules/Azure.AppConfig.AuditLogs/ Audit logs are required and I wasn't sure what to put for diagnostic log locations for the default testing. I left them all (Event Hub, Storage Account, Workspace). I felt this was heavy and not sure on guidance.

eriqua commented on avm/res/app-configuration/configuration-store/version.json:
```diff
@@ -1,7 +1,7 @@
 {
     "$schema": "https://aka.ms/bicep-registry-module-version-file-schema#",
-    "version": "0.1",
+    "version": "0.1.1",
```
Hey @JFolberth, no need to update the patch version, since that in Bicep is done automatically by the CI.

However, as per semver, given the addition of a child resource, and with that a new feature, I'd suggest bumping the minor version
```suggestion
    "version": "0.2",
```

eriqua commented on avm/res/app-configuration/configuration-store/tests/e2e/defaults/main.test.bicep:
```diff
@@ -43,6 +56,17 @@ module testDeployment '../../../main.bicep' = [
     params: {
       name: '${namePrefix}${serviceShort}001'
       location: resourceLocation
+      enablePurgeProtection: true
```
This can be a challenge in our testing, as we usually create an delete resources

JFolberth replied:
It is working in the test; however, the real failure is the order of operations as it feels the config store gets deleted first then the replica...which throws an error that the parent has already been deleted. I guess would ask how is this being accommodated for other resources who have purge protection such as Key Vault and Log Analytics?



I am not high on it being the default as it is not available in free tier and not sure on repeat logic as there already is logic to not allow if  `sku=='Free'` .I would suggest keeping it as false so to easily accommodate for more flexibility. 

`enablePurgeProtection: sku == 'Free' ? false : enablePurgeProtection`

eriqua replied:
I totally see the angle @JFolberth. And believe me a wormhole likely opens every time the discussion flexibility/usability vs security is raised 😄 
When it comes to WAF and especially reliability by default, that generally wins over flexibility. So the idea is to make the resource easier to deploy with WAF compliance than without.

However in this case I think we can achieve both. The implemented logic you mentioned should also work with the default value set to `true`, and that would be used by default unless the sku is set to free. What about adding more context in the enablePurgeProtection @description? Something like: `defaults to true unless 'sku' is set to 'Free', since purge protection is not available in Free tier.`

eriqua replied:
Unresolved this, since purge protection enablement in test files is currently blocking the validation pipeline. My suggestion is still the one of the first comment of this thread:

- Disable the feature in all tests and add a comment next to the assignment mentioning that is not WAF compliant but only an exception for deployment validation
- Add an exception for this rule in the psrule setting file, as done for keyvault for the same reason https://github.com/Azure/bicep-registry-modules/blob/main/avm/utilities/pipelines/staticValidation/psrule/ps-rule.yaml#L78

eriqua commented:
> @matebarabas @eriqua
> 
> Looking for feedback on this on two items.
> 
> I was not sure on replicas what to do with locations...I found avm\res\aad\domain-service\tests\e2e\waf-aligned\main.test.bicep where locations are hard coded and figured maybe this is how it should be handled?
> 
> I also per https://azure.github.io/PSRule.Rules.Azure/en/rules/Azure.AppConfig.AuditLogs/ Audit logs are required and I wasn't sure what to put for diagnostic log locations for the default testing. I left them all (Event Hub, Storage Account, Workspace). I felt this was heavy and not sure on guidance.

For the last point, those rules are currently not blocking as they are not part of the WAF reliability pillar. I'm not saying we shouldn't care about them, just that are not a priority 1 and can be taken for example in a next PR. 

That said, as a rule of thumb, defaults tests should only cover required parameters, so audit logs rules can be exempted for defaults tests only. The exemption file specific for defaults tests is https://github.com/Azure/bicep-registry-modules/blob/main/avm/utilities/pipelines/staticValidation/psrule/.ps-rule/min-suppress.Rule.yaml

eriqua commented on avm/res/app-configuration/configuration-store/replicas/main.bicep:
```diff
@@ -0,0 +1,31 @@
+metadata name = 'App Configuration Replica Configurations'
```
```suggestion
metadata name = 'App Configuration Replicas'
```

eriqua commented on avm/res/app-configuration/configuration-store/tests/e2e/defaults/main.test.bicep:
```diff
@@ -43,6 +45,9 @@ module testDeployment '../../../main.bicep' = [
       name: '${namePrefix}${serviceShort}001'
       location: resourceLocation
+      enablePurgeProtection: true
+      disableLocalAuth: true
+      replicaLocations: ['eastus', 'westus']
```
Defaults tests are supposed to only cover input parameters that are required. Ref https://azure.github.io/Azure-Verified-Modules/specs/bicep/#defaults-tests-must 

For this reason, I'd suggest to remove these optional parameters from this test, and add an exclusion to the defaults PSRule suppression group for the rule `Azure.AppConfig.GeoReplica`: https://github.com/Azure/bicep-registry-modules/blob/main/avm/utilities/pipelines/staticValidation/psrule/.ps-rule/min-suppress.Rule.yaml

Regarding `disableLocalAuth`, the PSRule Azure.AppConfig.DisableLocalAuth is not in the reliability pillar, thus non blocking for now. However, if we want to fix that already, I'd suggest to set its default value to `true` in the main.bicep template instead.

eriqua replied:
For replicas, yes, absolutely. Ideally we should also add them to the max test, which usually covers all possible features and child resources to maximize parameter validation.

JFolberth replied:
Done! I also had to add one to the min rule on Audit Logs.

eriqua commented on avm/res/app-configuration/configuration-store/tests/e2e/defaults/main.test.bicep:
```diff
@@ -24,6 +24,8 @@ param namePrefix string = '#_namePrefix_#'
 // Dependencies //
 // ============ //
 
+// Diagnostics
+// ===========
```
```suggestion
```

eriqua commented:
> > @matebarabas @eriqua
> > Looking for feedback on this on two items.
> > I was not sure on replicas what to do with locations...I found avm\res\aad\domain-service\tests\e2e\waf-aligned\main.test.bicep where locations are hard coded and figured maybe this is how it should be handled?
> > I also per https://azure.github.io/PSRule.Rules.Azure/en/rules/Azure.AppConfig.AuditLogs/ Audit logs are required and I wasn't sure what to put for diagnostic log locations for the default testing. I left them all (Event Hub, Storage Account, Workspace). I felt this was heavy and not sure on guidance.
> 
> For the last point, those rules are currently not blocking as they are not part of the WAF reliability pillar. I'm not saying we shouldn't care about them, just that are not a priority 1 and can be taken for example in a next PR.
> 
> That said, as a rule of thumb, defaults tests should only cover required parameters, so audit logs rules can be exempted for defaults tests only. The exemption file specific for defaults tests is https://github.com/Azure/bicep-registry-modules/blob/main/avm/utilities/pipelines/staticValidation/psrule/.ps-rule/min-suppress.Rule.yaml

@JFolberth now finally replying on the first point :)

The CI uses a region rotation for the deployment. So every test deployment may target a different location randomly picked up from [this list](https://github.com/Azure/bicep-registry-modules/blob/main/avm/utilities/pipelines/e2eValidation/regionSelector/Get-AzAvailableResourceLocation.ps1#L36): `eastus`, `uksouth`, `northeurope` and `eastasia`. 
That is unless a location is hardcoded in the test file, like the example you mentioned for AADDS. In this case, I guess we can rely on the default logic for the primary resource, while we should hardcode the location for the replica in the test, as you did already. 

I'm not an expert of App Store, but I'd expect it is not allowed to use the same location for the primary resource and the replica. So if, for example, the rotation logic picks 'eastus' for the primary resource, using 'eastus' for the replica would probably fail the deployment. Is that correct? If so,  for the hardcoded replica location we should pick a set not including the 4 locations we are rotating from.

JFolberth commented:
> > > @matebarabas @eriqua
> > > Looking for feedback on this on two items.
> > > I was not sure on replicas what to do with locations...I found avm\res\aad\domain-service\tests\e2e\waf-aligned\main.test.bicep where locations are hard coded and figured maybe this is how it should be handled?
> > > I also per https://azure.github.io/PSRule.Rules.Azure/en/rules/Azure.AppConfig.AuditLogs/ Audit logs are required and I wasn't sure what to put for diagnostic log locations for the default testing. I left them all (Event Hub, Storage Account, Workspace). I felt this was heavy and not sure on guidance.
> > 
> > 
> > For the last point, those rules are currently not blocking as they are not part of the WAF reliability pillar. I'm not saying we shouldn't care about them, just that are not a priority 1 and can be taken for example in a next PR.
> > That said, as a rule of thumb, defaults tests should only cover required parameters, so audit logs rules can be exempted for defaults tests only. The exemption file specific for defaults tests is https://github.com/Azure/bicep-registry-modules/blob/main/avm/utilities/pipelines/staticValidation/psrule/.ps-rule/min-suppress.Rule.yaml
> 
> @JFolberth now finally replying on the first point :)
> 
> The CI uses a region rotation for the deployment. So every test deployment may target a different location randomly picked up from [this list](https://github.com/Azure/bicep-registry-modules/blob/main/avm/utilities/pipelines/e2eValidation/regionSelector/Get-AzAvailableResourceLocation.ps1#L36): `eastus`, `uksouth`, `northeurope` and `eastasia`. That is unless a location is hardcoded in the test file, like the example you mentioned for AADDS. In this case, I guess we can rely on the default logic for the primary resource, while we should hardcode the location for the replica in the test, as you did already.
> 
> I'm not an expert of App Store, but I'd expect it is not allowed to use the same location for the primary resource and the replica. So if, for example, the rotation logic picks 'eastus' for the primary resource, using 'eastus' for the replica would probably fail the deployment. Is that correct? If so, for the hardcoded replica location we should pick a set not including the 4 locations we are rotating from.

I am not overly familiar with it app stores as well. I have updated. My thought was with two to 1.) ensure we hit a region that the resource is not deployed to and 2.) ensure it can handle multiple locations. I have just updated it not to use `eastus` so this should be accomplished.

eriqua commented on avm/res/app-configuration/configuration-store/tests/e2e/max/main.test.bicep:
```diff
@@ -67,6 +67,7 @@ module testDeployment '../../../main.bicep' = [
       name: '${namePrefix}${serviceShort}001'
       location: resourceLocation
       createMode: 'Default'
+      replicaLocations: ['eastus', 'westus']
```
```suggestion
      replicaLocations: ['centralus', 'westus']
```

eriqua commented:
> > > > @matebarabas @eriqua
> > > > Looking for feedback on this on two items.
> > > > I was not sure on replicas what to do with locations...I found avm\res\aad\domain-service\tests\e2e\waf-aligned\main.test.bicep where locations are hard coded and figured maybe this is how it should be handled?
> > > > I also per https://azure.github.io/PSRule.Rules.Azure/en/rules/Azure.AppConfig.AuditLogs/ Audit logs are required and I wasn't sure what to put for diagnostic log locations for the default testing. I left them all (Event Hub, Storage Account, Workspace). I felt this was heavy and not sure on guidance.
> > > 
> > > 
> > > For the last point, those rules are currently not blocking as they are not part of the WAF reliability pillar. I'm not saying we shouldn't care about them, just that are not a priority 1 and can be taken for example in a next PR.
> > > That said, as a rule of thumb, defaults tests should only cover required parameters, so audit logs rules can be exempted for defaults tests only. The exemption file specific for defaults tests is https://github.com/Azure/bicep-registry-modules/blob/main/avm/utilities/pipelines/staticValidation/psrule/.ps-rule/min-suppress.Rule.yaml
> > 
> > 
> > @JFolberth now finally replying on the first point :)
> > The CI uses a region rotation for the deployment. So every test deployment may target a different location randomly picked up from [this list](https://github.com/Azure/bicep-registry-modules/blob/main/avm/utilities/pipelines/e2eValidation/regionSelector/Get-AzAvailableResourceLocation.ps1#L36): `eastus`, `uksouth`, `northeurope` and `eastasia`. That is unless a location is hardcoded in the test file, like the example you mentioned for AADDS. In this case, I guess we can rely on the default logic for the primary resource, while we should hardcode the location for the replica in the test, as you did already.
> > I'm not an expert of App Store, but I'd expect it is not allowed to use the same location for the primary resource and the replica. So if, for example, the rotation logic picks 'eastus' for the primary resource, using 'eastus' for the replica would probably fail the deployment. Is that correct? If so, for the hardcoded replica location we should pick a set not including the 4 locations we are rotating from.
> 
> I am not overly familiar with app stores as well. I have updated. My thought was with two to 1.) ensure we hit a region that the resource is not deployed to and 2.) ensure it can handle multiple locations. I have just updated it not to use `westus` and `centralus` so this should be accomplished.

Sounds great. Just added a suggestion for the other test file to do the same. Overall it looks good. As a final step, could you please 
- Rerun the set-AVMModule utility to update the Readmes
- Retrigger the pipeline manually to verify if the run works after enabling purge protection by default. Based on the result I may need to look into the CI deployment removal logic.

Thanks!

JFolberth commented:
> > > > > @matebarabas @eriqua
> > > > > Looking for feedback on this on two items.
> > > > > I was not sure on replicas what to do with locations...I found avm\res\aad\domain-service\tests\e2e\waf-aligned\main.test.bicep where locations are hard coded and figured maybe this is how it should be handled?
> > > > > I also per https://azure.github.io/PSRule.Rules.Azure/en/rules/Azure.AppConfig.AuditLogs/ Audit logs are required and I wasn't sure what to put for diagnostic log locations for the default testing. I left them all (Event Hub, Storage Account, Workspace). I felt this was heavy and not sure on guidance.
> > > > 
> > > > 
> > > > For the last point, those rules are currently not blocking as they are not part of the WAF reliability pillar. I'm not saying we shouldn't care about them, just that are not a priority 1 and can be taken for example in a next PR.
> > > > That said, as a rule of thumb, defaults tests should only cover required parameters, so audit logs rules can be exempted for defaults tests only. The exemption file specific for defaults tests is https://github.com/Azure/bicep-registry-modules/blob/main/avm/utilities/pipelines/staticValidation/psrule/.ps-rule/min-suppress.Rule.yaml
> > > 
> > > 
> > > @JFolberth now finally replying on the first point :)
> > > The CI uses a region rotation for the deployment. So every test deployment may target a different location randomly picked up from [this list](https://github.com/Azure/bicep-registry-modules/blob/main/avm/utilities/pipelines/e2eValidation/regionSelector/Get-AzAvailableResourceLocation.ps1#L36): `eastus`, `uksouth`, `northeurope` and `eastasia`. That is unless a location is hardcoded in the test file, like the example you mentioned for AADDS. In this case, I guess we can rely on the default logic for the primary resource, while we should hardcode the location for the replica in the test, as you did already.
> > > I'm not an expert of App Store, but I'd expect it is not allowed to use the same location for the primary resource and the replica. So if, for example, the rotation logic picks 'eastus' for the primary resource, using 'eastus' for the replica would probably fail the deployment. Is that correct? If so, for the hardcoded replica location we should pick a set not including the 4 locations we are rotating from.
> > 
> > 
> > I am not overly familiar with app stores as well. I have updated. My thought was with two to 1.) ensure we hit a region that the resource is not deployed to and 2.) ensure it can handle multiple locations. I have just updated it not to use `westus` and `centralus` so this should be accomplished.
> 
> Sounds great. Just added a suggestion for the other test file to do the same. Overall it looks good. As a final step, could you please
> 
> * Rerun the set-AVMModule utility to update the Readmes
> * Retrigger the pipeline manually to verify if the run works after enabling purge protection by default. Based on the result I may need to look into the CI deployment removal logic.
> 
> Thanks!

All updated!

eriqua commented on avm/res/app-configuration/configuration-store/main.bicep:
```diff
@@ -26,10 +26,10 @@ param sku string = 'Standard'
 
-@description('Optional. Property specifying whether protection against purge is enabled for this configuration store.')
-param enablePurgeProtection bool = false
+@description('Optional. Property specifying whether protection against purge is enabled for this configuration store.Defaults to true unless sku is set to Free, since purge protection is not available in Free tier.')
```
Sorry, just noticed this minor one
```suggestion
@description('Optional. Property specifying whether protection against purge is enabled for this configuration store. Defaults to true unless sku is set to Free, since purge protection is not available in Free tier.')
```

JFolberth replied:
Thanks @eriqua I have committed this change.

eriqua replied:
Hey @JFolberth thanks. We are very close! Please run the Set-AVMModule utility to reflect the change in the main.json as well. Also please rerun the pipeline so that it can hopefully have a green badge

eriqua reviewed (approved):
🆗
